src/sage/modular/modform/element.py

```diff
@@ -233,9 +265,111 @@ def __call__(self, x, prec=None):
+            sage: f.qexp(3000).polynomial()(ComplexField(1024)(1+2*I)/3)  # rel tol 1e-300
+            429.1999483220629427868808539905635963271200790650662541094321031444732211998294807859026647488684939077694414494735552293875615266296656556614279454721135728033076949678382364730635806027464404827172930052286699587458446255120255079481396375697779539931587961853350866816125810074340461757133364142825540647 - 786.1573628418824335115383082485297499477188534110060762218331702495288763372243559079588574592451375320461099650989125359884247389330529818992955559828420621763451597667207389282918920501399439528649925730381086098314012196716268053157349502456867311327972354510982957748538682030274690668695122466979462069*I
+
+        Check ``SR`` does not make the result lose precision::
+
+            sage: f(ComplexField(1024)(1+2*I)/3 + x - x)  # rel tol 1e-300
+            429.199948322062942786880853990563596327120079065066254109432103144473221199829480785902664748868493907769441449473555229387561526629665655661427945472113572803307694967838236473063580602746440482717293005228669958745844625512025507948139637569777953993158796185335086681612581007434046175713336414282554064647719988594288913120 - 786.157362841882433511538308248529749947718853411006076221833170249528876337224355907958857459245137532046109965098912535988424738933052981899295555982842062176345159766720738928291892050139943952864992573038108609831401219671626805315734950245686731132797235451098295774853868203027469066869512246697946206845252244686704009880*I
+        """
+        from sage.rings.integer import Integer
+        from sage.misc.functional import log
+        from sage.structure.element import parent
+        from sage.rings.complex_mpfr import ComplexNumber
+        from sage.rings.cc import CC
+        from sage.rings.real_mpfr import RealNumber
+        from sage.symbolic.constants import pi
+        from sage.rings.imaginary_unit import I  # import from here instead of sage.symbolic.constants to avoid cast to SR
+        from sage.symbolic.expression import Expression
+        if isinstance(x, Expression):
+            try:
+                x = x.pyobject()
+            except TypeError:
+                pass
+        if x in CC:
+            if x == 0:
+                return self.qexp(1)[0]
+            if not isinstance(x, (RealNumber, ComplexNumber)):
+                x = CC(x)  # might lose precision if this is done unconditionally (TODO what about interval and ball types?)
+        if isinstance(x, (RealNumber, ComplexNumber)):
+            return self.eval_at_tau(log(x)/(2*parent(x)(pi)*I))  # cast to parent(x) to force numerical evaluation of pi
         return self.q_expansion(prec)(x)
 
+    def eval_at_tau(self, tau):
+        r"""
+        Evaluate this modular form at the half-period ratio `\tau`.
+        This is related to `q` by `q = e^{2\pi i \tau}`.
+
+        TESTS:
```

Comment:
Could we have at least one public doctest (with an `EXAMPLES::` block) since it is a public method?